src/lib/operators/table_scan.cpp

```diff
@@ -0,0 +1,220 @@
+    if (comparator_function(value)) {
+      const auto chunk_offset = (*ref_seg_position_list)[segment_index];
+      pos_list.emplace_back(chunk_offset);
+    }
+  }
+}
+
+template <typename T>
+void TableScan::_scan_segment(const ChunkID& chunk_id, std::shared_ptr<BaseSegment>& segment, PosList& pos_list,
+                              std::function<bool(const T)>& comparator_function, const T& typed_search_value) {
+  auto reference_segment = std::dynamic_pointer_cast<ReferenceSegment>(segment);
+  if (reference_segment) {
+    _scan_reference_segment(chunk_id, *reference_segment, pos_list, comparator_function);
+    return;
+  }
+
+  auto dictionary_segment = std::dynamic_pointer_cast<DictionarySegment<T>>(segment);
+  if (dictionary_segment) {
+    _scan_dictionary_segment(chunk_id, *dictionary_segment, pos_list, typed_search_value);
+    return;
+  }
+
+  auto value_segment = std::dynamic_pointer_cast<ValueSegment<T>>(segment);
+  if (value_segment) {
+    _scan_value_segment(chunk_id, *value_segment, pos_list, comparator_function);
+    return;
+  }
+
+  throw std::runtime_error("There is no segment scan implemented for this segment type.");
+}
+
+template <typename T>
+T TableScan::_get_typed_search_value() {
+  try {
+    return type_cast<T>(_search_value);
+  } catch (...) {
+    throw std::invalid_argument("Target column type and compare value type are incompatible.");
+  }
+}
+
+std::shared_ptr<const Table> TableScan::_on_execute() {
+  const auto input_table = _input_operator->get_output();
+  const auto column_count = input_table->column_count();
+
+  auto output_table = std::make_shared<Table>();
+  for (auto column_id = ColumnID{0}; column_id < column_count; ++column_id) {
+    output_table->copy_column_definition(input_table, column_id);
+  }
+
+  if (input_table->is_empty()) {
+    return output_table;
+  }
+
+  const auto column_type_name = input_table->column_type(_column_id);
+  resolve_data_type(column_type_name, [&](auto type) {
+    using Type = typename decltype(type)::type;
+    auto typed_search_value = _get_typed_search_value<Type>();
+    auto comparator_function = _build_comparator_function(typed_search_value);
+
+    auto reference_position_list = std::make_shared<PosList>();
+
+    // scan input segments
+    const auto chunk_count = input_table->chunk_count();
+    for (auto chunk_id = ChunkID{0}; chunk_id < chunk_count; ++chunk_id) {
+      auto segment = input_table->get_chunk(chunk_id).get_segment(_column_id);
+      _scan_segment<Type>(chunk_id, segment, *reference_position_list, comparator_function, typed_search_value);
+    }
+
+    auto first_reference_segment =
+        std::dynamic_pointer_cast<ReferenceSegment>(input_table->get_chunk(ChunkID{0}).get_segment(ColumnID{0}));
+    // check if input_table contains reference segments, if true we have to get the original table from it
+    const auto referenced_table = (first_reference_segment) ? first_reference_segment->referenced_table() : input_table;
```

Comment:
This might be a valid criticism which we also thought about, but we come to the conclusion that this holds true for the general case and not only the current implementation. Our point of view comes from the assignment documentation:
```
There is one shortcoming of this approach: Because a ReferenceSegment cannot point 
to values in different tables (e.g., because two tables were concatenated using a union), 
we could not use the same approach for the sort operator if the first chunk was part of 
a different table. In this case, we will resort to materializing the inputs. Since this problem 
only happens with the sort operator, which is usually the last operator in a query plan, 
this materialization is not harmful for performance, as it would have to occur anyway.
```
All "reference tables" (tables which result from operators - therefore only having reference chunks) shall reference to the same original table, otherwise they have been materialized and do not contain reference chunks.